added docs for header auth strategy (#340)
kiali/kiali.io: content/documentation/staging/configuration/authentication/header.adoc

relref "../rbac" >}}[Role-based access control documentation] for more details.

== Set-up

The `header` strategy will work with any Kubernetes cluster.  The token provided 
must be supported by that cluster.  For instance, most "on-prem" clusters support
OpenID Connect, but cloud hosted clusters do not.  For clusters that don't support 
a token, the link:https://kubernetes.io/docs/reference/access-authn-authz/authentication/#user-impersonation[impersonation] 
headers can be injected by the reverse proxy.

[source,yaml]
----
spec:
  auth:
    strategy: header
----

The `header` strategy doesn't have any additional configuration.

== HTTP Header

The `header` strategy looks for a token in the `Authorization` HTTP header with the 
`Bearer` prefix.  The HTTP header should look like:

[source]
----
Authorization: Bearer TOKEN
----

Where `TOKEN` is the appropriate token for your cluster.  This `TOKEN` will be 
submitted to the API server via a `TokenReview` to validate the token *ONLY*
on the first access to Kiali.  On subsequent calls the `TOKEN` is passed through
directly to the API server.

== Security Considerations

> === Network Polcies

Reviewer: Fix to typo

```suggestion
=== Network Policies
```

> A policy should be put in place to make sure that the only "client" for kiali is

Reviewer 2: ```suggestion
A policy should be put in place to make sure that the only "client" for Kiali is
```

the authenticating reverse proxy.  This helps limit potential abuse and ensures 
that the authenticating reverse proxy is the source of truth for who accessed

> kiali.

Reviewer 2: ```suggestion
Kiali.
```

=== Short Lived Tokens

> The authenticating reverse proxy should inject short lived token in the 

Reviewer 2: ```suggestion
The authenticating reverse proxy should inject a short lived token in the 
```

`Authorization` header.  A shorter lived token is less likely to be abused if 
leaked.  Kiali will take whatever token is passed into the reqeuest, so as tokens 
are regenerated Kiali will use the new token.

=== Impersonation

==== TokenRequest API

The authenticating reverse proxy should use the TokenRequest API instead of static 
`ServiceAccount` tokens when possible while using impersonation. The 
`ServiceAccount` that can impersonate users and groups is privileged and having it
be short lived cuts down on the possibility of a token being leaked while it's being 
passed between different parts of the infrastructure.

> ==== Drop Incomming Impersonation Headers

Reviewer: Another typo

```suggestion
==== Drop Incoming Impersonation Headers
```

The authenticating proxy *MUST* drop any headers it receives from a remote client 

> that mach the impersonation headers.  Not only do you want to make sure that the

Reviewer 2: ```suggestion
that match the impersonation headers.  Not only do you want to make sure that the
```

> authenticating proxy can't be over-riden on which user to authenticate, but also

Reviewer 2: ```suggestion
authenticating proxy can't be overriden on which user to authenticate, but also
```
